File: core/src/main/java/io/github/msameer0/rhythmicrush/screens/ui/HudRenderer.java
package io.github.msameer0.rhythmicrush.screens.ui;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.graphics.OrthographicCamera;
import com.badlogic.gdx.graphics.g2d.BitmapFont;
import com.badlogic.gdx.graphics.g2d.GlyphLayout;
import com.badlogic.gdx.graphics.g2d.SpriteBatch;
import com.badlogic.gdx.graphics.glutils.ShapeRenderer;
import com.badlogic.gdx.utils.Align;
import com.badlogic.gdx.utils.viewport.Viewport;

import io.github.msameer0.rhythmicrush.RhythmicRushGame;
import io.github.msameer0.rhythmicrush.game.GameWorld;
import io.github.msameer0.rhythmicrush.game.level.LevelProgress;
import io.github.msameer0.rhythmicrush.settings.SettingsManager;

/**
 * Draws all in-game HUD elements: progress bar, attempt/best/FPS counters,
 * the pause button, and the "New Best" popup.
 *
 * <p>Owned by {@code GameScreen} and called from its {@code draw()} method.
 * Requires an active SpriteBatch / ShapeRenderer context to be managed by the
 * caller — see individual method docs for which renderer must be open.</p>
 */
public class HudRenderer {

    // ── Popup timing ──────────────────────────────────────────────────────────
    private static final float POPUP_FADE_IN  = 0.25f;
    private static final float POPUP_HOLD     = 1.20f;
    private static final float POPUP_FADE_OUT = 0.45f;
    private static final float POPUP_TOTAL    = POPUP_FADE_IN + POPUP_HOLD + POPUP_FADE_OUT;

    private static final float PAUSE_BTN = 44f;

    // ── Colours ───────────────────────────────────────────────────────────────
    private static final Color COL_FILL    = new Color(0.35f, 0.65f, 1.00f, 1f);
    private static final Color COL_HEADING = new Color(1f, 0.85f, 0.35f, 1f);
    private static final Color HUD_ATTEMPT = new Color(1f, 1f, 1f, 0.85f);
    private static final Color HUD_BEST    = new Color(1f, 1f, 1f, 0.55f);
    private static final Color HUD_FPS     = new Color(1f, 1f, 1f, 0.45f);

    // ── Dependencies ──────────────────────────────────────────────────────────
    private final RhythmicRushGame game;
    private final GameWorld world;
    private final BitmapFont font;
    private final GlyphLayout glyphLayout;
    private final ShapeRenderer shapes;
    private final SpriteBatch batch;

    // ── State (owned by GameScreen but read here) ─────────────────────────────
    private final StringBuilder sb = new StringBuilder(32);

    private float popupTimer  = -1f;
    private int   popupBestPct = 0;

    public HudRenderer(RhythmicRushGame game, GameWorld world,
                       BitmapFont font, ShapeRenderer shapes, SpriteBatch batch) {
        this.game       = game;
        this.world      = world;
        this.font       = font;
        this.glyphLayout = new GlyphLayout();
        this.shapes     = shapes;
        this.batch      = batch;
    }

    // ── Update ────────────────────────────────────────────────────────────────

    /**
     * Advances the "New Best" popup timer. Call once per frame even when paused.
     *
     * @param delta Seconds since last frame.
     */
    public void update(float delta) {
        if (popupTimer >= 0f) {
            popupTimer += delta;
            if (popupTimer >= POPUP_TOTAL) popupTimer = -1f;
        }
    }

    /** Triggers the "New Best" popup animation. */
    public void showNewBestPopup(int bestPct) {
        popupTimer    = 0f;
        popupBestPct  = bestPct;
    }

    // ── Shape draws (call inside a ShapeRenderer.Filled block) ───────────────

    /**
     * Draws the progress bar background and fill shapes.
     * Must be called inside an open {@code shapes.begin(Filled)} block.
     */
    public void drawProgressBarShapes(OrthographicCamera camera, Viewport viewport) {
        float progress = world.getProgress();
        if (progress <= 0f) return;
        SettingsManager s = game.getSettingsManager();
        if (!s.getShowProgressBar()) return;

        final float BAR_W = viewport.getWorldWidth() * 0.625f * 0.55f;
        final float BAR_H = 10f;
        final float GAP   = 14f;
        final float LINE_Y = camTop(camera, viewport) - (s.getUiPadding() + 6f);

        float textW = 0f;
        if (s.getShowPercentage()) {
            sb.setLength(0);
            sb.append(Math.round(progress * 100f)).append('%');
            font.getData().setScale(1.2f);
            glyphLayout.setText(font, sb, Color.WHITE, 0, Align.left, false);
            textW = glyphLayout.width;
        }

        float totalW  = (s.getShowPercentage() ? textW + GAP : 0f) + BAR_W;
        float startX  = camera.position.x - totalW / 2f;
        float r       = BAR_H / 2f;
        float fillW   = BAR_W * progress;

        shapes.setColor(0.2f, 0.2f, 0.2f, 0.55f);
        drawRoundedRect(startX, LINE_Y - BAR_H / 2f, BAR_W, BAR_H, r);

        if (fillW >= BAR_H) {
            shapes.setColor(COL_FILL);
            drawRoundedRect(startX, LINE_Y - BAR_H / 2f, fillW, BAR_H, r);
        } else if (fillW > 0) {
            shapes.setColor(COL_FILL);
            shapes.rect(startX, LINE_Y - BAR_H / 2f, fillW, BAR_H);
        }
    }

    /**
     * Draws the pause button circle and bar icons.
     * Must be called inside an open {@code shapes.begin(Filled)} block.
     */
    public void drawPauseButtonShapes(OrthographicCamera camera, Viewport viewport) {
        float cx = pauseCircleCX(camera, viewport);
        float cy = pauseCircleCY(camera, viewport);
        float r  = PAUSE_BTN / 2f;

        shapes.setColor(0.2f, 0.2f, 0.2f, 0.75f);
        shapes.circle(cx, cy, r, 32);

        float barW = r * 0.22f, barH = r * 0.75f, gap = r * 0.18f;
        shapes.setColor(1f, 1f, 1f, 0.9f);
        shapes.rect(cx - gap - barW, cy - barH / 2f, barW, barH);
        shapes.rect(cx + gap,        cy - barH / 2f, barW, barH);
    }

    // ── Text draws (call inside a SpriteBatch.begin block) ────────────────────

    /**
     * Draws the percentage label beside the progress bar.
     * Must be called inside an open {@code batch.begin()} block.
     */
    public void drawProgressBarText(OrthographicCamera camera, Viewport viewport, String levelKey) {
        float progress = world.getProgress();
        if (progress <= 0f) return;
        SettingsManager s = game.getSettingsManager();
        if (!s.getShowPercentage()) return;

        int pct = Math.round(progress * 100f);
        final float BAR_W  = viewport.getWorldWidth() * 0.625f * 0.55f;
        final float GAP    = 14f;
        final float LINE_Y = camTop(camera, viewport) - (s.getUiPadding() + 6f);

        sb.setLength(0);
        sb.append(pct).append('%');
        font.getData().setScale(1.2f);
        glyphLayout.setText(font, sb, Color.WHITE, 0, Align.left, false);
        float textW = glyphLayout.width;
        float textH = glyphLayout.height;

        float totalW    = textW + (s.getShowProgressBar() ? GAP + BAR_W : 0f);
        float startX    = camera.position.x - totalW / 2f;
        float textDrawX = s.getShowProgressBar() ? startX + BAR_W + GAP : startX;

        boolean isPB = false;
        if (levelKey != null) {
            LevelProgress p = game.getProgressManager().getOrCreate(levelKey);
            isPB = pct > p.getBestPercent();
        }
        Color textColor = isPB ? COL_HEADING : Color.WHITE;

        font.setColor(0, 0, 0, textColor.a * 0.4f);
        font.draw(batch, sb, textDrawX + 2f, LINE_Y + textH / 2f - 2f);
        font.setColor(textColor);
        font.draw(batch, sb, textDrawX, LINE_Y + textH / 2f);
        font.getData().setScale(1f);
    }

    /**
     * Draws attempts, best %, and optional FPS in the top-left corner.
     * Must be called inside an open {@code batch.begin()} block.
     */
    public void drawSessionAttemptsText(OrthographicCamera camera, Viewport viewport,
                                        int sessionAttempts, String levelKey) {
        SettingsManager s = game.getSettingsManager();
        float p    = s.getUiPadding();
        float left = camLeft(camera, viewport) + p;
        float top  = camTop(camera, viewport) - p;
        float shadowOffset = 2f;
        float nextY = top;

        if (s.getShowAttempts()) {
            sb.setLength(0);
            sb.append("Attempt  ").append(sessionAttempts);
            font.setColor(0, 0, 0, HUD_ATTEMPT.a * 0.4f);
            font.draw(batch, sb, left + shadowOffset, nextY - shadowOffset);
            font.setColor(HUD_ATTEMPT);
            font.draw(batch, sb, left, nextY);
            nextY -= 26f;
        }

        if (s.getShowBest() && levelKey != null) {
            LevelProgress lp = game.getProgressManager().getOrCreate(levelKey);
            sb.setLength(0);
            sb.append("Best  ").append(lp.getBestPercent()).append('%');
            font.setColor(0, 0, 0, HUD_BEST.a * 0.4f);
            font.draw(batch, sb, left + shadowOffset, nextY - shadowOffset);
            font.setColor(HUD_BEST);
            font.draw(batch, sb, left, nextY);
            nextY -= 26f;
        }

        if (s.getShowFps()) {
            sb.setLength(0);
            sb.append("FPS  ").append(Gdx.graphics.getFramesPerSecond());
            font.setColor(0, 0, 0, HUD_FPS.a * 0.4f);
            font.draw(batch, sb, left + shadowOffset, nextY - shadowOffset);
            font.setColor(HUD_FPS);
            font.draw(batch, sb, left, nextY);
        }
    }

    /**
     * Draws the animated "New Best" popup.
     * Must be called inside an open {@code batch.begin()} block.
     */
    public void drawNewBestPopup(OrthographicCamera camera) {
        if (popupTimer < 0f) return;

        float alpha, scale;
        if (popupTimer < POPUP_FADE_IN) {
            float t = popupTimer / POPUP_FADE_IN;
            alpha = t;
            scale = 1.0f + 0.8f * t;
        } else if (popupTimer < POPUP_FADE_IN + POPUP_HOLD) {
            float t = (popupTimer - POPUP_FADE_IN) / POPUP_HOLD;
            alpha = 1f;
            scale = 1.8f - 0.4f * t;
        } else {
            float t = (popupTimer - POPUP_FADE_IN - POPUP_HOLD) / POPUP_FADE_OUT;
            alpha = 1f - t;
            scale = 1.4f * (1f - t);
        }
        alpha = Math.max(0f, Math.min(1f, alpha));
        scale = Math.max(0f, scale);

        float cx = camera.position.x;
        float cy = camera.position.y;

        font.getData().setScale(scale);
        sb.setLength(0);
        sb.append("NEW BEST");
        glyphLayout.setText(font, sb);
        float textH = glyphLayout.height;
        float textY = cy + textH / 2f;

        font.setColor(0, 0, 0, alpha * 0.4f);
        font.draw(batch, sb, cx - glyphLayout.width / 2f + 2f, textY - 2f);
        font.setColor(COL_HEADING.r, COL_HEADING.g, COL_HEADING.b, alpha);
        font.draw(batch, sb, cx - glyphLayout.width / 2f, textY);

        font.getData().setScale(scale * 0.6f);
        sb.setLength(0);
        sb.append(popupBestPct).append('%');
        glyphLayout.setText(font, sb);
        float pctY = textY - textH - 5f;

        font.setColor(0, 0, 0, alpha * 0.85f * 0.4f);
        font.draw(batch, sb, cx - glyphLayout.width / 2f + 2f, pctY - 2f);
        font.setColor(1f, 1f, 1f, alpha * 0.85f);
        font.draw(batch, sb, cx - glyphLayout.width / 2f, pctY);

        font.getData().setScale(1f);
        font.setColor(Color.WHITE);
    }

    // ── Hit testing ───────────────────────────────────────────────────────────

    /**
     * Returns whether a world-space coordinate lands on the pause button.
     *
     * @param tx World X of the touch.
     * @param ty World Y of the touch.
     */
    public boolean hitsPauseButton(float tx, float ty, OrthographicCamera camera, Viewport viewport) {
        float cx = pauseCircleCX(camera, viewport);
        float cy = pauseCircleCY(camera, viewport);
        float r  = PAUSE_BTN / 2f + 8f;
        float dx = tx - cx, dy = ty - cy;
        return dx * dx + dy * dy <= r * r;
    }

    // ── Camera convenience ────────────────────────────────────────────────────

    private static float camTop(OrthographicCamera camera, Viewport viewport) {
        return camera.position.y + viewport.getWorldHeight() / 2f;
    }

    private static float camLeft(OrthographicCamera camera, Viewport viewport) {
        return camera.position.x - viewport.getWorldWidth() / 2f;
    }

    private float pauseCircleCX(OrthographicCamera camera, Viewport viewport) {
        return camera.position.x + viewport.getWorldWidth() / 2f
            - PAUSE_BTN / 2f - (game.getSettingsManager().getUiPadding() + 2f);
    }

    private float pauseCircleCY(OrthographicCamera camera, Viewport viewport) {
        return camera.position.y + viewport.getWorldHeight() / 2f
            - PAUSE_BTN / 2f - (game.getSettingsManager().getUiPadding() + 2f);
    }

    // ── Shape helper ──────────────────────────────────────────────────────────

    private void drawRoundedRect(float x, float y, float w, float h, float r) {
        shapes.rect(x + r, y,     w - 2 * r, h);
        shapes.rect(x,     y + r, r,          h - 2 * r);
        shapes.rect(x + w - r, y + r, r,      h - 2 * r);
        shapes.circle(x + r,     y + r,     r, 16);
        shapes.circle(x + w - r, y + r,     r, 16);
        shapes.circle(x + r,     y + h - r, r, 16);
        shapes.circle(x + w - r, y + h - r, r, 16);
    }
}

File: core/src/main/kotlin/io/github/msameer0/rhythmicrush/screens/ui/HudRenderer.kt
package io.github.msameer0.rhythmicrush.screens.ui

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.graphics.OrthographicCamera
import com.badlogic.gdx.graphics.g2d.BitmapFont
import com.badlogic.gdx.graphics.g2d.GlyphLayout
import com.badlogic.gdx.graphics.g2d.SpriteBatch
import com.badlogic.gdx.graphics.glutils.ShapeRenderer
import com.badlogic.gdx.math.MathUtils
import com.badlogic.gdx.utils.Align
import com.badlogic.gdx.utils.viewport.Viewport
import io.github.msameer0.rhythmicrush.RhythmicRushGame
import io.github.msameer0.rhythmicrush.game.GameWorld

/**
 * Handles the rendering of the gameplay HUD, including the progress bar, attempts, and UI popups.
 */
class HudRenderer(
    private val game: RhythmicRushGame,
    private val world: GameWorld,
    private val font: BitmapFont,
    private val shapes: ShapeRenderer,
    private val batch: SpriteBatch
) {

    companion object {
        private const val POPUP_FADE_IN = 0.25f
        private const val POPUP_HOLD = 1.20f
        private const val POPUP_FADE_OUT = 0.45f
        private const val POPUP_TOTAL = POPUP_FADE_IN + POPUP_HOLD + POPUP_FADE_OUT

        private const val PAUSE_BTN = 44f

        private val COL_FILL = Color(0.35f, 0.65f, 1.00f, 1f)
        private val COL_HEADING = Color(1f, 0.85f, 0.35f, 1f)
        private val HUD_ATTEMPT = Color(1f, 1f, 1f, 0.85f)
        private val HUD_BEST = Color(1f, 1f, 1f, 0.55f)
        private val HUD_FPS = Color(1f, 1f, 1f, 0.45f)
        private val PROGRESS_TRACK = Color(0.2f, 0.2f, 0.2f, 0.55f)
        private val PAUSE_BUTTON = Color(0.2f, 0.2f, 0.2f, 0.75f)
        private val PAUSE_ICON = Color(1f, 1f, 1f, 0.9f)
    }

    private val glyphLayout = GlyphLayout()
    private val textBuilder = StringBuilder(32)

    private var popupTimer = -1f
    private var popupBestPct = 0

    fun update(delta: Float) {
        if (popupTimer >= 0f) {
            popupTimer += delta
            if (popupTimer >= POPUP_TOTAL) popupTimer = -1f
        }
    }

    fun showNewBestPopup(bestPct: Int) {
        popupTimer = 0f
        popupBestPct = bestPct
    }

    fun hideNewBestPopup() {
        popupTimer = -1f
    }

    fun drawProgressBarShapes(camera: OrthographicCamera, viewport: Viewport) {
        val progress = world.progress
        if (progress <= 0f) return
        val settings = game.settingsManager
        if (!settings.showProgressBar) return

        val barW = viewport.worldWidth * 0.625f * 0.55f
        val barH = 10f
        val gap = 14f
        val lineY = camTop(camera, viewport) - (settings.uiPadding + 6f)

        var textW = 0f
        if (settings.showPercentage) {
            textBuilder.setLength(0)
            textBuilder.append(MathUtils.round(progress * 100f)).append('%')
            font.data.setScale(1.2f)
            glyphLayout.setText(font, textBuilder, Color.WHITE, 0f, Align.left, false)
            textW = glyphLayout.width
        }

        val totalW = (if (settings.showPercentage) textW + gap else 0f) + barW
        val startX = camera.position.x - totalW / 2f
        val cornerRadius = barH / 2f
        val fillW = barW * progress

        shapes.color = PROGRESS_TRACK
        drawRoundedRect(startX, lineY - barH / 2f, barW, barH, cornerRadius)

        if (fillW >= barH) {
            shapes.color = COL_FILL
            drawRoundedRect(startX, lineY - barH / 2f, fillW, barH, cornerRadius)
        } else if (fillW > 0) {
            shapes.color = COL_FILL
            shapes.rect(startX, lineY - barH / 2f, fillW, barH)
        }
    }

    fun drawPauseButtonShapes(camera: OrthographicCamera, viewport: Viewport) {
        val centerX = pauseCircleCX(camera, viewport)
        val centerY = pauseCircleCY(camera, viewport)
        val radius = PAUSE_BTN / 2f

        shapes.color = PAUSE_BUTTON
        shapes.circle(centerX, centerY, radius, 32)

        val barWidth = radius * 0.22f
        val barHeight = radius * 0.75f
        val gap = radius * 0.18f
        shapes.color = PAUSE_ICON
        shapes.rect(
            centerX - gap - barWidth,
            centerY - barHeight / 2f,
            barWidth,
            barHeight
        )
        shapes.rect(centerX + gap, centerY - barHeight / 2f, barWidth, barHeight)
    }

    fun drawProgressBarText(camera: OrthographicCamera, viewport: Viewport, levelKey: String?) {
        val progress = world.progress
        if (progress <= 0f) return
        val settings = game.settingsManager
        if (!settings.showPercentage) return

        val pct = MathUtils.round(progress * 100f)
        val barW = viewport.worldWidth * 0.625f * 0.55f
        val gap = 14f
        val lineY = camTop(camera, viewport) - (settings.uiPadding + 6f)

        textBuilder.setLength(0)
        textBuilder.append(pct).append('%')
        font.data.setScale(1.2f)
        glyphLayout.setText(font, textBuilder, Color.WHITE, 0f, Align.left, false)
        val textW = glyphLayout.width
        val textH = glyphLayout.height

        val totalW = textW + (if (settings.showProgressBar) gap + barW else 0f)
        val startX = camera.position.x - totalW / 2f
        val textDrawX =
            if (settings.showProgressBar) startX + barW + gap else startX

        var isPersonalBest = false
        if (levelKey != null) {
            val levelProgress = game.progressManager.getOrCreate(levelKey)
            isPersonalBest = pct > levelProgress.bestPercent
        }
        val textColor = if (isPersonalBest) COL_HEADING else Color.WHITE

        font.setColor(0f, 0f, 0f, textColor.a * 0.4f)
        font.draw(batch, textBuilder, textDrawX + 2f, lineY + textH / 2f - 2f)
        font.color = textColor
        font.draw(batch, textBuilder, textDrawX, lineY + textH / 2f)
        font.data.setScale(1f)
    }

    fun drawSessionAttemptsText(
        camera: OrthographicCamera, viewport: Viewport,
        sessionAttempts: Int, levelKey: String?
    ) {
        val settings = game.settingsManager
        val padding = settings.uiPadding
        val left = camLeft(camera, viewport) + padding
        val top = camTop(camera, viewport) - padding
        val shadowOffset = 2f
        var nextY = top

        if (settings.showAttempts) {
            textBuilder.setLength(0)
            textBuilder.append("Attempt  ").append(sessionAttempts)
            font.setColor(0f, 0f, 0f, HUD_ATTEMPT.a * 0.4f)
            font.draw(batch, textBuilder, left + shadowOffset, nextY - shadowOffset)
            font.color = HUD_ATTEMPT
            font.draw(batch, textBuilder, left, nextY)
            nextY -= 26f
        }

        if (settings.showBest && levelKey != null) {
            val levelProgress = game.progressManager.getOrCreate(levelKey)
            textBuilder.setLength(0)
            textBuilder.append("Best  ").append(levelProgress.bestPercent).append('%')
            font.setColor(0f, 0f, 0f, HUD_BEST.a * 0.4f)
            font.draw(batch, textBuilder, left + shadowOffset, nextY - shadowOffset)
            font.color = HUD_BEST
            font.draw(batch, textBuilder, left, nextY)
            nextY -= 26f
        }

        if (settings.showFps) {
            textBuilder.setLength(0)
            textBuilder.append("FPS  ").append(Gdx.graphics.framesPerSecond)
            font.setColor(0f, 0f, 0f, HUD_FPS.a * 0.4f)
            font.draw(batch, textBuilder, left + shadowOffset, nextY - shadowOffset)
            font.color = HUD_FPS
            font.draw(batch, textBuilder, left, nextY)
        }
    }

    fun drawNewBestPopup(camera: OrthographicCamera) {
        if (popupTimer < 0f) return

        var alpha: Float
        var scale: Float
        if (popupTimer < POPUP_FADE_IN) {
            val fadeProgress = popupTimer / POPUP_FADE_IN
            alpha = fadeProgress
            scale = 1.0f + 0.8f * fadeProgress
        } else if (popupTimer < POPUP_FADE_IN + POPUP_HOLD) {
            val holdProgress = (popupTimer - POPUP_FADE_IN) / POPUP_HOLD
            alpha = 1f
            scale = 1.8f - 0.4f * holdProgress
        } else {
            val fadeProgress =
                (popupTimer - POPUP_FADE_IN - POPUP_HOLD) / POPUP_FADE_OUT
            alpha = 1f - fadeProgress
            scale = 1.4f * (1f - fadeProgress)
        }
        alpha = alpha.coerceIn(0f, 1f)
        scale = scale.coerceAtLeast(0f)

        val centerX = camera.position.x
        val centerY = camera.position.y

        font.data.setScale(scale)
        textBuilder.setLength(0)
        textBuilder.append("NEW BEST")
        glyphLayout.setText(font, textBuilder)
        val textH = glyphLayout.height
        val textY = centerY + textH / 2f

        font.setColor(0f, 0f, 0f, alpha * 0.4f)
        font.draw(
            batch,
            textBuilder,
            centerX - glyphLayout.width / 2f + 2f,
            textY - 2f
        )
        font.setColor(COL_HEADING.r, COL_HEADING.g, COL_HEADING.b, alpha)
        font.draw(batch, textBuilder, centerX - glyphLayout.width / 2f, textY)

        font.data.setScale(scale * 0.6f)
        textBuilder.setLength(0)
        textBuilder.append(popupBestPct).append('%')
        glyphLayout.setText(font, textBuilder)
        val pctY = textY - textH - 5f

        font.setColor(0f, 0f, 0f, alpha * 0.85f * 0.4f)
        font.draw(
            batch,
            textBuilder,
            centerX - glyphLayout.width / 2f + 2f,
            pctY - 2f
        )
        font.setColor(1f, 1f, 1f, alpha * 0.85f)
        font.draw(batch, textBuilder, centerX - glyphLayout.width / 2f, pctY)

        font.data.setScale(1f)
        font.color = Color.WHITE
    }

    fun hitsPauseButton(
        touchX: Float,
        touchY: Float,
        camera: OrthographicCamera,
        viewport: Viewport
    ): Boolean {
        val centerX = pauseCircleCX(camera, viewport)
        val centerY = pauseCircleCY(camera, viewport)
        val radius = PAUSE_BTN / 2f + 8f
        val offsetX = touchX - centerX
        val offsetY = touchY - centerY
        return offsetX * offsetX + offsetY * offsetY <= radius * radius
    }

    private fun camTop(camera: OrthographicCamera, viewport: Viewport): Float {
        return camera.position.y + viewport.worldHeight / 2f
    }

    private fun camLeft(camera: OrthographicCamera, viewport: Viewport): Float {
        return camera.position.x - viewport.worldWidth / 2f
    }

    private fun pauseCircleCX(camera: OrthographicCamera, viewport: Viewport): Float {
        return camera.position.x + viewport.worldWidth / 2f - PAUSE_BTN / 2f - (game.settingsManager.uiPadding + 2f)
    }

    private fun pauseCircleCY(camera: OrthographicCamera, viewport: Viewport): Float {
        return camera.position.y + viewport.worldHeight / 2f - PAUSE_BTN / 2f - (game.settingsManager.uiPadding + 2f)
    }

    private fun drawRoundedRect(x: Float, y: Float, w: Float, h: Float, r: Float) {
        shapes.rect(x + r, y, w - 2 * r, h)
        shapes.rect(x, y + r, r, h - 2 * r)
        shapes.rect(x + w - r, y + r, r, h - 2 * r)
        shapes.circle(x + r, y + r, r, 16)
        shapes.circle(x + w - r, y + r, r, 16)
        shapes.circle(x + r, y + h - r, r, 16)
        shapes.circle(x + w - r, y + h - r, r, 16)
    }
}
